### PR TITLE
LogMemberLeaveHandler

### LINKED ISSUE #101: Member Leave Embed
When `guildMemberRemove` is triggered, an embed should be sent in the channel `LOG_CHANNEL_ID` that sends an embed stating:
- How long the member was in the guild (years, days, hours, minutes and seconds)
- If the user was authenticated (they will have the role `MEMBER_ROLE` if so)

I am not sure if you can get the roles a user had from the quit event, so may need some investigation.

### PR DESCRIPTION
**Pull request by BlackBearFTW**
This pull request adds a handler for when a member leaves, it will send a message to the logs channel that states:
- When the user joined and left.
- How long the user stayed on the server.
- if the user was authenticated.

Besides that, it also adds a new `DateUtil` method called `getFormattedTimeSinceDate`:
- Takes in a `startDate: Date` and a `endDate: Date` as argument.
- returns a string that is formated as followed `# years, # days, # hours, # minutes and # seconds`.
- will not return certain parts of the string if value is `0`.
- will detect if output value is single `day` or multiple `days`.
- will return null when `endDate` is earlier then `startDate`.

Knipsel

### CODE REVIEW

File: src/event/handlers/LogMemberLeaveHandler.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+class LogMemberLeaveHandler extends EventHandler {
+	constructor() {
+		super(Constants.Events.GUILD_MEMBER_REMOVE);
+	}
+
+	async handle(guildMember: GuildMember): Promise <void> {
+		const embed = new MessageEmbed();
+
+		embed.setTitle(`${guildMember.user.tag} left the server.`);
```
Comment (LamboCreeper) [inline src/event/handlers/LogMemberLeaveHandler.ts]: For consistency with the other logs, can the title please be "Member Left" with the description of embed being "User: `@mention`" (you should still be able to mention users who are no longer in the guild).

File: src/event/handlers/LogMemberLeaveHandler.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+		embed.setColor(EMBED_COLOURS.DEFAULT);
+		embed.addField("Join date", new Date(guildMember.joinedTimestamp!).toLocaleString(), true);
+		embed.addField("Leave date", new Date(Date.now()).toLocaleString(), true);
+		embed.addField("Time in server", DateUtils.getFormattedTimeSinceDate(guildMember.joinedTimestamp!, Date.now()));
```
Comment (LamboCreeper) [inline src/event/handlers/LogMemberLeaveHandler.ts]: Each word in the title of these fields should begin with a capital letter as it is a heading.

File: src/utils/DateUtils.ts [hunk 3]
```diff
@@ -20,6 +20,41 @@ class DateUtils {
 
 		return `${time} on ${date.getDate()} ${month} ${date.getFullYear()}`;
 	}
+
+	static getFormattedTimeSinceDate(startTimeMs: number, endTimeMs: number): string | null {
+		if (endTimeMs < startTimeMs) return null;
+
+		let difference = endTimeMs - startTimeMs;
+
+		let daysDifference = Math.floor(difference / 1000 / 60 / 60 / 24);
```
Comment (LamboCreeper) [inline src/utils/DateUtils.ts]: Remove the blank line between these please.

File: src/utils/DateUtils.ts [hunk 4]
```diff
@@ -20,6 +20,41 @@ class DateUtils {
...
+
+		difference -= hoursDifference * 1000 * 60 * 60;
+
+		const minutesDifference = Math.floor(difference / 1000 / 60);
+
+		difference -= minutesDifference * 1000 * 60;
+
+		const secondDifference = Math.floor(difference / 1000);
+
+		const yearsDifference = Math.floor(daysDifference / 365);
```
Comment (LamboCreeper) [inline src/utils/DateUtils.ts]: Remove the blank lines between these please.

File: src/event/handlers/LogMemberLeaveHandler.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import EventHandler from "../../abstracts/EventHandler";
+import { EMBED_COLOURS, MEMBER_ROLE, LOG_CHANNEL_ID } from "../../config.json";
+import DateUtils from "../../utils/DateUtils";
+
+class LogMemberLeaveHandler extends EventHandler {
+	constructor() {
+		super(Constants.Events.GUILD_MEMBER_REMOVE);
+	}
+
+	async handle(guildMember: GuildMember): Promise <void> {
```
Comment (LamboCreeper) [inline src/event/handlers/LogMemberLeaveHandler.ts]: Should not be a space before the generic.

File: test/utils/DateUtilsTest.ts [hunk 6]
```diff
@@ -69,4 +69,14 @@ describe("DateUtils", () => {
...
 		});
 	});
+
+	describe("::getFormattedTimeSinceDate()", () => {
+		it("returns a formatted string that contains the difference between two dates", () => {
+			expect(DateUtils.getFormattedTimeSinceDate(new Date("01/12/2021 00:00:00").getTime(), new Date("01/15/2022 06:30:05").getTime())).to.equal("1 year, 3 days, 6 hours, 30 minutes and 5 seconds");
+		});
+	});
+
+	it("returns null if endTimeMs is earlier then startTimeMs", () => {
```
Comment (LamboCreeper) [inline test/utils/DateUtilsTest.ts]: This `it()` does not belong to the `describe()`.

File: test/utils/DateUtilsTest.ts [hunk 7]
```diff
@@ -69,4 +69,14 @@ describe("DateUtils", () => {
 			expect(DateUtils.formatAsText(date)).to.equal("09:05 on 7 Feb 2010");
 		});
 	});
+
+	describe("::getFormattedTimeSinceDate()", () => {
+		it("returns a formatted string that contains the difference between two dates", () => {
+			expect(DateUtils.getFormattedTimeSinceDate(new Date("01/12/2021 00:00:00").getTime(), new Date("01/15/2022 06:30:05").getTime())).to.equal("1 year, 3 days, 6 hours, 30 minutes and 5 seconds");
```
Comment (LamboCreeper) [inline test/utils/DateUtilsTest.ts]: This line is rather long, can we break this up a bit. i.e:
```ts
const expected = "1 year, 3 days, 6 hours, 30 minutes and 5 seconds";
const start = new Date("01/12/2021 00:00:00").getTime();
const end = new Date("01/15/2022 06:30:05").getTime();
const actual = DateUtils.getFormattedTimeSinceDate(start, end);

expect(actual).to.equal(expected);

File: src/utils/DateUtils.ts [hunk 8]
```diff
@@ -20,6 +20,41 @@ class DateUtils {
 
 		return `${time} on ${date.getDate()} ${month} ${date.getFullYear()}`;
 	}
+
+	static getFormattedTimeSinceDate(startTimeMs: number, endTimeMs: number): string | null {
```
Comment (LamboCreeper) [inline src/utils/DateUtils.ts]: I believe I asked this to accept two dates, not numbers.

**Review by LamboCreeper — CHANGES_REQUESTED** [review]

File: test/utils/DateUtilsTest.ts (hunk 6 above)
Comment (BlackBearFTW) [reply test/utils/DateUtilsTest.ts]: Where does it belong?

File: src/utils/DateUtils.ts [hunk 9]
```diff
@@ -20,6 +20,36 @@ class DateUtils {
...
 	}
+
+	static getFormattedTimeSinceDate(startDate: Date, endDate: Date): string | null {
+		if (endDate.getTime() < startDate.getTime()) return null;
+
+		let difference = endDate.getTime() - startDate.getTime();
+		let daysDifference = Math.floor(difference / 1000 / 60 / 60 / 24);
+
+		difference -= daysDifference * 1000 * 60 * 60 * 24;
+		const hoursDifference = Math.floor(difference / 1000 / 60 / 60);
```
Comment (LamboCreeper) [inline src/utils/DateUtils.ts]: Add a blank line between these please.

File: src/utils/DateUtils.ts [hunk 10]
```diff
@@ -20,6 +20,36 @@ class DateUtils {
...
+
+		difference -= hoursDifference * 1000 * 60 * 60;
+		const minutesDifference = Math.floor(difference / 1000 / 60);
```
Comment (LamboCreeper) [inline src/utils/DateUtils.ts]: Add a blank line between these please.

File: src/utils/DateUtils.ts [hunk 11]
```diff
@@ -20,6 +20,36 @@ class DateUtils {
...
+
+		difference -= minutesDifference * 1000 * 60;
+		const secondDifference = Math.floor(difference / 1000);
```
Comment (LamboCreeper) [inline src/utils/DateUtils.ts]: Add a blank line between these please.

**Review by LamboCreeper — CHANGES_REQUESTED** [review]
Sorry, I must have missed these before. All good to go after that.

**Review by LamboCreeper — APPROVED** [review]

**Pull request merged**